src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -1438,185 +1449,169 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
-                            &proj.base,
-                            buf,
-                            autoderef,
-                            &including_downcast,
-                        )?;
-                        buf.push_str(&"[..]");
+                }
+                ProjectionElem::Downcast(..) => {
+                    if including_downcast.0 {
+                        return Err(());
                     }
-                };
+                }
+                ProjectionElem::Field(field, _ty) => {
+                    let field = if place_str.is_empty() {
+                        self.mir.upvar_decls[field.index()].debug_name.to_string()
+                    } else {
+                        let prefix = place.prefix(self.infcx.tcx, i);
+                        let field_name = self.describe_field(&prefix, *field);
+                        format!(".{}", field_name)
+                    };
+                    place_str.push_str(field.as_str());
+                }
+                ProjectionElem::Index(index) => {
+                    let index = if let Ok(index) = self.local_str(*index) {
+                        index
+                    } else {
+                        "_".to_string()
+                    };
+                    place_str.push_str(format!("[{}]", index).as_str());
+                }
+                ProjectionElem::ConstantIndex { .. }
+                | ProjectionElem::Subslice { .. } => {
+                    // Since it isn't possible to borrow an element on a particular index and
+                    // then use another while the borrow is held, don't output indices details
+                    // to avoid confusing the end-user
+                    place_str.push_str("[..]");
+                }
             }
         }
 
-        Ok(())
+        Ok(place_str)
     }
 
     /// Appends end-user visible description of the `local` place to `buf`. If `local` doesn't have
     /// a name, then `Err` is returned
-    fn append_local_to_string(&self, local_index: Local, buf: &mut String) -> Result<(), ()> {
+    fn local_str(&self, local_index: Local) -> Result<String, ()> {
         let local = &self.mir.local_decls[local_index];
-        match local.name {
-            Some(name) => {
-                buf.push_str(&name.to_string());
-                Ok(())
-            }
-            None => Err(()),
+        debug!("describe_place-base-local: {:?}", local);
+
+        if let Some(name) = local.name {
+            Ok(name.to_string())
+        } else {
+            Err(())
         }
     }
 
     /// End-user visible description of the `field`nth field of `base`
-    fn describe_field(&self, base: &Place, field: Field) -> String {
-        match *base {
-            Place::Local(local) => {
-                let local = &self.mir.local_decls[local];
+    fn describe_field(&self, base: &NeoPlace<'tcx>, field: Field) -> String {
+        let mut buf = match &base.base {
```

Comment:
this logic doesn't seem right. This is yet another case where the existing code, anyway, is processing "from the right-hand side", but this logic is written to process "from the beginning". In other words, if `base` is `a.b.c`, the old code would process the `c` first (and stop) -- but this code processes `a` then `b` then `c`. Seems like it will make strange strings, if I'm reading the code correctly, and even if not it just feels more complicated.

Reply:
as you can see later in the `for` loop, this `mut buf` will be replaced by elems `c` and just return(stop), not appended as `a.b.c`.